R/similarModel.R

```diff
@@ -0,0 +1,122 @@
+library(methods)
+library(parallel)
+
+#' S4 class that represents similar products recommendation model
+#' @export
+#' @param sim cosine matrix of product similarity
+setClass("similarity.recommender", representation(sim = "matrix"))
+
+#' Factory for the similarity recommendation model
+#' @export
+#' @param data product hit stream with visitor.id and sku
+#' @param filter allows to reduce recommendations to a given set
+similarityRecommender <- function(data, filter = NULL) {
+  m <- userProductHitsToMatrix(data)
+  m <- cosineMatrix(m)
+  if(is.null(filter)) filter <- colnames(m)
+  m <- m[, filter]
+  model <- new("similarity.recommender", sim = m)
+  return(model)
+}
+
+
+#' Makes similarity score predictions based on the similarity.recommender model
+#' @param object a fitted similarity model object.
+#' @param newdata use product hits.
+#' @return \code{predict} returns a vector with predicted values.
+#' @rdname similarity
+#' @aliases predict,similarity-recommender
+#' @export
+setMethod("predict", signature(object = "similarity.recommender"),
+          function(object, newdata) {
+            similarity.predictor(object, newdata)
+          })
+
+
+#' Predicts similarity score for new product hits data
+#'
+#' @importFrom stats predict
+#' @export
+#' @param object similarity model object
+#' @param newdata product hits data
+similarity.predictor <- function(object, newdata) {
+  sku <- score <- NULL
+  colnames(newdata) <- c("visitor.id", "sku", "sku.rec")
+
+  target.skus <- intersect(unique(newdata[, sku]), colnames(object@sim))
```

Comment:
This is covered by test:

```R
  # SKU does not exist in the matrix:
  page.views <-  data.table(visitor.id = c("u1"), sku = c("z"))
  res <- recommendSimilarProducts(test.sim.model, page.views)
  expect_identical(nrow(res), 0L, "Result is empty for the SKU that is not in the similarity model")
```

Here: https://github.com/byapparov/recommender/blob/predict/tests/testthat/test.similarity.model.R